homeassistant/helpers/icon.py

```diff
@@ -1,7 +1,165 @@
 """Icon helper methods."""
 from __future__ import annotations
 
+import asyncio
+from collections.abc import Iterable
 from functools import lru_cache
+import logging
+from typing import Any
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.loader import Integration, async_get_integrations
+from homeassistant.util.json import load_json_object
+
+from .translation import build_resources
+
+ICON_LOAD_LOCK = "icon_load_lock"
+ICON_CACHE = "icon_cache"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def _component_icons_path(component: str, integration: Integration) -> str | None:
+    """Return the icons json file location for a component.
+
+    Ex: components/hue/icons.json
+    If component is just a single file, will return None.
+    """
+    domain = component.rpartition(".")[-1]
+
+    # If it's a component that is just one file, we don't support icons
+    # Example custom_components/my_component.py
+    if integration.file_path.name != domain:
+        return None
+
+    return str(integration.file_path / "icons.json")
+
+
+def _load_icons_files(icons_files: dict[str, str]) -> dict[str, dict[str, Any]]:
+    """Load and parse icons.json files."""
+    return {
+        component: load_json_object(icons_file)
+        for component, icons_file in icons_files.items()
+    }
+
+
+async def _async_get_component_icons(
+    hass: HomeAssistant,
+    components: set[str],
+    integrations: dict[str, Integration],
+) -> dict[str, Any]:
+    """Load icons."""
+    icons: dict[str, Any] = {}
+
+    # Determine files to load
+    files_to_load = {}
+    for loaded in components:
+        domain = loaded.rpartition(".")[-1]
+        if (path := _component_icons_path(loaded, integrations[domain])) is None:
+            icons[loaded] = {}
+        else:
+            files_to_load[loaded] = path
+
+    # Load files
+    if files_to_load and (
+        load_icons_job := hass.async_add_executor_job(_load_icons_files, files_to_load)
```

Comment:
`load_icons_job` will never be falsy, as it will always return a Future. I think we can remove the second condition.